File: Level 7 - System Design & Probabilistic DS/t_digest/advanced_problems.java
import java.util.*;

/**
 * Advanced T-Digest Problems in Java
 * Focuses on merging multiple independent T-Digests.
 */
public class advanced_problems {

    static class SimpleTDigest {
        List<basic_operations.Centroid> centroids = new ArrayList<>();
        double totalWeight = 0;

        void add(double val) {
            centroids.add(new basic_operations.Centroid(val, 1.0));
            totalWeight++;
        }

        void merge(SimpleTDigest other) {
            this.centroids.addAll(other.centroids);
            this.totalWeight += other.totalWeight;
            // Re-compression would happen here in a real impl
        }

        double getQuantile(double q) {
            Collections.sort(centroids);
            double target = q * totalWeight;
            double current = 0;
            for (basic_operations.Centroid c : centroids) {
                if (current + c.weight >= target)
                    return c.mean;
                current += c.weight;
            }
            return centroids.get(centroids.size() - 1).mean;
        }
    }

    public static void main(String[] args) {
        SimpleTDigest regionA = new SimpleTDigest();
        SimpleTDigest regionB = new SimpleTDigest();

        // Simulate two different data streams
        for (int i = 0; i < 5000; i++)
            regionA.add(i);
        for (int i = 5000; i < 10000; i++)
            regionB.add(i);

        System.out.println("Region A P50: " + regionA.getQuantile(0.5)); // ~2500
        System.out.println("Region B P50: " + regionB.getQuantile(0.5)); // ~7500

        System.out.println("\nMerging Region A and Region B...");
        regionA.merge(regionB);
        System.out.println("Global P50: " + regionA.getQuantile(0.5)); // ~5000
        System.out.println("Global P99: " + regionA.getQuantile(0.99)); // ~9900
    }
}

File: Level 7 - System Design & Probabilistic DS/t_digest/advanced_problems.cpp
#include <iostream>
#include <vector>
#include <algorithm>

/**
 * Advanced T-Digest Problems in C++
 * Focuses on merging digests.
 */

struct Centroid {
    double mean, weight;
    bool operator<(const Centroid& o) const { return mean < o.mean; }
};

class SimpleTDigest {
public:
    std::vector<Centroid> centroids;
    double totalWeight = 0;

    void add(double v) { centroids.push_back({v, 1.0}); totalWeight++; }
    
    void merge(const SimpleTDigest& other) {
        centroids.insert(centroids.end(), other.centroids.begin(), other.centroids.end());
        totalWeight += other.totalWeight;
    }

    double quantile(double q) {
        if (centroids.empty()) return 0;
        std::sort(centroids.begin(), centroids.end());
        double target = q * totalWeight;
        double current = 0;
        for (const auto& c : centroids) {
            if (current + c.weight >= target) return c.mean;
            current += c.weight;
        }
        return centroids.back().mean;
    }
};

int main() {
    SimpleTDigest s1, s2;
    for(int i=0; i<1000; ++i) s1.add(i);
    for(int i=1000; i<2000; ++i) s2.add(i);

    std::cout << "S1 P50: " << s1.quantile(0.5) << std::endl;
    std::cout << "S2 P50: " << s2.quantile(0.5) << std::endl;

    s1.merge(s2);
    std::cout << "Merged P50: " << s1.quantile(0.5) << " (Expected: 1000)" << std::endl;

    return 0;
}
